pennylane/devices/default_mixed.py

```diff
@@ -0,0 +1,280 @@
+        kraus_index = ABC[
+            2 * (self.num_wires + len(channel_wires)) : 2 * (self.num_wires + len(channel_wires))
+            + 1
+        ]
+
+        # new state indices replace row and column indices with new ones
+        new_state_indices = functools.reduce(
+            lambda old_string, idx_pair: old_string.replace(idx_pair[0], idx_pair[1]),
+            zip(col_indices + row_indices, new_col_indices + new_row_indices),
+            state_indices,
+        )
+
+        # index mapping for einsum, e.g., 'iga,abcdef,idh->gbchef'
+        einsum_indices = (
+            "{kraus_index}{new_row_indices}{row_indices}, {state_indices}"
+            "{kraus_index}{col_indices}{new_col_indices}->{new_state_indices}".format(
+                kraus_index=kraus_index,
+                new_col_indices=new_col_indices,
+                col_indices=col_indices,
+                state_indices=state_indices,
+                row_indices=row_indices,
+                new_row_indices=new_row_indices,
+                new_state_indices=new_state_indices,
+            )
+        )
+
+        self._state = self._einsum(einsum_indices, kraus, self._state, kraus_dagger)
+
+    def _apply_diagonal_unitary(self, eigvals, wires):
+        r"""Apply a diagonal unitary gate specified by a list of eigenvalues. This method uses
+        the fact that the unitary is diagonal for a more efficient implementation.
+
+        Args:
+            eigvals (array): eigenvalues (phases) of the diagonal unitary
+            wires (Wires): target wires
+        """
+
+        channel_wires = self.map_wires(wires)
+
+        # Tensor indices of the state. For each qubit, need an index for rows *and* columns
+        state_indices = ABC[: 2 * self.num_wires]
+
+        # row indices of the quantum state affected by this operation
+        row_wires_list = channel_wires.tolist()
+        row_indices = "".join(ABC_ARRAY[row_wires_list].tolist())
+
+        # column indices are shifted by the number of wires
+        col_wires_list = [w + self.num_wires for w in row_wires_list]
+        col_indices = "".join(ABC_ARRAY[col_wires_list].tolist())
+
+        einsum_indices = "{row_indices},{state_indices},{col_indices}->{state_indices}".format(
+            col_indices=col_indices, state_indices=state_indices, row_indices=row_indices
+        )
+
+        self._state = self._einsum(einsum_indices, eigvals, self._state, self._conj(eigvals))
+
+    def _apply_operation(self, operation):
+        """Applies operations to the internal device state.
+
+        Args:
+            operation (~.Operation): operation to apply on the device
+        """
+        wires = operation.wires
+        matrices = self._get_kraus(operation)
+
+        if isinstance(operation, DiagonalOperation):
+            self._apply_diagonal_unitary(matrices, wires)
+        else:
+            self._apply_channel(matrices, wires)
+
+    def apply(self, operations, rotations=None, **kwargs):
```

Comment:
(Just a heads-up: this signature might be up for change soon. Fortunately, the new signature would be fairly similar: it would include a composite object (`CircuitGraph` or `QuantumTape`). Both `operations` and `rotations` would be extractable from this object, so the overall change would be relatively minor.)